### ISSUE TITLE
0xjoi - [H-1] "Attackers can fabricate balances for ERC20 tokens that don't yet exist, enabling them to set traps that could steal funds from users who interact with these tokens in the future."

### ISSUE THREAD

**Issue by sherlock-admin2** [sherlock-audit/2024-07-sense-points-marketplace-judging#178]
0xjoi

High

# [H-1] "Attackers can fabricate balances for ERC20 tokens that don't yet exist, enabling them to set traps that could steal funds from users who interact with these tokens in the future."

Fake balances can be created for not-yet-existing ERC20 tokens, which allows attackers to set traps to steal funds from future users

## **Location:** [File name and line numbers]

- https://github.com/sherlock-audit/2024-07-sense-points-marketplace/blob/main/point-tokenization-vault/contracts/PointTokenVault.sol#L114
- https://github.com/sherlock-audit/2024-07-sense-points-marketplace/blob/main/point-tokenization-vault/contracts/PointTokenVault.sol#L130

## **Description**

there is a subtle difference between the implementation of solmate's SafeTransferLib and OZ's SafeERC20:

OZ's SafeERC20 checks if the token is a contract or not, solmate's SafeTransferLib does not.

See: https://github.com/Rari-Capital/solmate/blob/main/src/utils/SafeTransferLib.sol#L9

Note that none of the functions in this library check that a token has code at all! That responsibility is delegated to the caller.

As a result, when the token's address has no code, the transaction will just succeed with no error.

This attack vector was made well-known by the qBridge hack back in Jan 2022.

However, it's becoming popular for protocols to deploy their token across multiple networks and when they do so, a common practice is to deploy the token contract from the same deployer address and with the same nonce so that the token address can be the same for all the networks.

For example: $1INCH is using the same token address for both Ethereum and BSC; Gelato's $GEL token is using the same token address for Ethereum, Fantom and Polygon.

A sophisticated attacker can exploit it by taking advantage of that and setting traps on multiple potential tokens to steal from the future users that deposits with such tokens.

## **Impact**

Stealing of deposited funds from users

## **Proof of Concepts**

Given:

ProjectA has TokenA on another network;
ProjectB has TokenB on another network;
ProjectC has TokenC on another network;
The attacker deposit() for TokenA, TokenB, and TokenC with 10000e18 each;
A few months later, ProjectB lunched TokenB on the local network at the same address;
Alice deposit() with 11000e18 TokenB;
The attacker called withdraw() to receive 10000e18 TokenB.
In summary, one of the traps set by the attacker was activated by the deployment of TokenB and Alice was the victim. As a result, 10000e18 TokenB was stolen by the attacker.

## **Tools Used**

Manual Review

## **Recommended mitigation**

Consider using OZ's SafeERC20 instead.
